### PR TITLE
fix: improve parent typing and fix repeat context type bug

### LINKED ISSUE #6396: fix: Repeat Directive's Type Signature is incorrect
# 🐛 Bug Report
It appears the type signature for the `repeat` directive is incorrect. Specifically, the acceptance of `Expression<TSource, TArray, ExecutionContext<TSource>>` types. This type expression forces the `context.parent` property of the expression binding to be of type `ExecutionContext<TSource>`, which is inaccurate. See a repro below, along with the difference in type signatures between the `when` directive and the `repeat` directive.

It's also worth noting that in the `when` directive, the `context.parent` property is of type `any`, which isn't accurate. @EisenbergEffect I recall you saying these types get messy quickly and getting everything to flow through properly was very tricky, any idea what is going on here? I'm noticing that `CaptureType` doesn't accept a type param for `TParent`, and that many directives don't capture that information - should they?

## 💻 Repro or Code Sample
https://stackblitz.com/edit/typescript-vcy276?file=index.ts

### PR DESCRIPTION
**Pull request by EisenbergEffect**
# Pull Request

## 📖 Description

In various type refactorings it looks like `repeat`'s types got messed up a bit. This PR fixes that as well as tries to plumb through parent types more.

### 🎫 Issues

* Closes #6396 

## 👩‍💻 Reviewer Notes

No functional changes. Just fixing types.

## 📑 Test Plan

All tests still pass.

### General

- [x] I have included a change request file using `$ yarn change`
- [ ] I have added tests for my changes.
- [x] I have tested my changes.
- [ ] I have updated the project documentation to reflect my changes.
- [x] I have read the [CONTRIBUTING](https://github.com/microsoft/fast/blob/master/CONTRIBUTING.md) documentation and followed the [standards](/docs/community/code-of-conduct/#our-standards) for this project.

## ⏭ Next Steps

Still exploring improving inference in other areas, such as repeat item templates. It may not be possible though. I'm speaking with the TS team to see if they have any ideas. If we find a solution, I'll submit an additional PR for that.

### CODE REVIEW

File: packages/web-components/fast-element/src/templating/repeat.ts [hunk 1]
```diff
@@ -19,6 +18,7 @@ import {
 import type { CaptureType, SyntheticViewTemplate, ViewTemplate } from "./template.js";
 import { HTMLView, SyntheticView } from "./view.js";
 import { normalizeBinding } from "./binding.js";
+// import { when } from "./when.js";
```
Comment (janechu) [inline packages/web-components/fast-element/src/templating/repeat.ts:21]: should we just delete this?

Comment (nicholasrice) [inline packages/web-components/fast-element/src/templating/repeat.ts:21]:
```suggestion
```

**Review by nicholasrice — APPROVED** [review]

File: packages/web-components/fast-element/src/templating/repeat.ts (hunk 1 above)
Comment (EisenbergEffect) [reply packages/web-components/fast-element/src/templating/repeat.ts:21]: Oops. That was from some testing. Will fix. One sec.

Comment (EisenbergEffect) [reply packages/web-components/fast-element/src/templating/repeat.ts:21]: On it.

**Review by janechu — APPROVED** [review]

**Pull request merged**